convert_hf_to_gguf.py

```diff
@@ -7581,6 +7682,132 @@ def modify_tensors(self, data_torch: Tensor, name: str, bid: int | None) -> Iter
+            weights.reshape(
+                n_head, 2, weights.shape[0] // n_head // 2, *weights.shape[1:]
+            )
+            .swapaxes(1, 2)
+            .reshape(weights.shape)
+        )
+
+    def modify_tensors(
+        self, data_torch: Tensor, name: str, bid: int | None
+    ) -> Iterable[tuple[str, Tensor]]:
+        n_head = self.hparams["n_heads"]
+        n_kv_head = self.hparams.get("n_kv_heads")
+        is_vision_tensor = any(
+            name.startswith(prefix)
+            for prefix in [
+                "vision_encoder.",
+                "vision_language_adapter.",
+                "patch_merger.",
+                "pre_mm_projector_norm",
+            ]
+        )
+
+        if is_vision_tensor:
+            return []  # skip vision tensors
+
+        if self.undo_permute:
+            if name.endswith(("q_proj.weight", "q_proj.bias")):
+                data_torch = self.permute(data_torch, n_head, n_head)
+            if name.endswith(("k_proj.weight", "k_proj.bias")):
+                data_torch = self.permute(data_torch, n_head, n_kv_head)
+
+        return [(self.map_tensor_name(name), data_torch)]
+
+
+class PixtralModel(MmprojModel):
+    model_name = "Pixtral"
+    img_break_tok_id = -1
+
+    def __init__(self, *args, **kwargs):
+        super().__init__(*args, **kwargs)
+        # layer_norm_eps is not in config.json, it is hard-coded in modeling_pixtral.py
+        self.hparams["layer_norm_eps"] = self.hparams.get("norm_eps", 1e-5)
+        self.img_break_tok_id = self.find_vparam(["image_break_token_id"])
+        logger.info(f"Image break token id: {self.img_break_tok_id}")
+
+    def set_gguf_parameters(self):
+        super().set_gguf_parameters()
+        self.gguf_writer.add_clip_projector_type(gguf.VisionProjectorType.PIXTRAL)
+
+        self.gguf_writer.add_vision_attention_layernorm_eps(
+            self.find_hparam(["layer_norm_eps"])
+        )
+        self.gguf_writer.add_rope_freq_base(self.find_vparam(["rope_theta"]))
+
+        self.gguf_writer.add_vision_use_silu(True)
+
+        # spatial_merge_size
+        if self.find_vparam(["mm_projector_id"]) == "patch_merge":
+            self.gguf_writer.add_vision_spatial_merge_size(
+                self.find_vparam(["spatial_merge_size"])
+            )
+
+    def modify_tensors(
+        self, data_torch: Tensor, name: str, bid: int | None
+    ) -> Iterable[tuple[str, Tensor]]:
+        del bid  # unused
+        n_head = self.find_vparam(["num_attention_heads"])
+        n_kv_head = n_head
+
+        if any(
+            name.startswith(prefix)
+            for prefix in [
+                "vision_encoder.",
+                "vision_language_adapter.",
+                "patch_merger.",
+                "pre_mm_projector_norm",
+            ]
+        ):
+            # process vision tensors
+            if name.endswith(("q_proj.weight", "q_proj.bias")):
+                data_torch = MistralModel.permute(data_torch, n_head, n_head)
```

Comment:
IIRC the original model is not permuted, only the HF model is.

So having this line could lead to a broken model